Frends.HTTP.DownloadFile/Frends.HTTP.DownloadFile/Extensions.cs

```diff
@@ -0,0 +1,129 @@
+
+        if (options.AllowInvalidCertificate)
+            handler.ServerCertificateCustomValidationCallback = (a, b, c, d) => true;
+    }
+
+    internal static void SetDefaultRequestHeadersBasedOnOptions(this HttpClient httpClient, Options options)
+    {
+        //Do not automatically set expect 100-continue response header
+        httpClient.DefaultRequestHeaders.ExpectContinue = false;
+        httpClient.DefaultRequestHeaders.TryAddWithoutValidation("content-type", "application/json");
+        httpClient.Timeout = TimeSpan.FromSeconds(Convert.ToDouble(options.ConnectionTimeoutSeconds));
+    }
+
+    private static X509Certificate[] GetCertificates(Options options)
+    {
+        X509Certificate2[] certificates;
+
+        switch (options.ClientCertificateSource)
+        {
+            case CertificateSource.CertificateStore:
+                var thumbprint = options.CertificateThumbprint;
+                certificates = GetCertificatesFromStore(thumbprint, options.LoadEntireChainForCertificate);
+                break;
+            case CertificateSource.File:
+                certificates = GetCertificatesFromFile(options.ClientCertificateFilePath, options.ClientCertificateKeyPhrase);
+                break;
+            case CertificateSource.String:
+                certificates = GetCertificatesFromString(options.ClientCertificateInBase64, options.ClientCertificateKeyPhrase);
+                break;
+            default:
+                throw new Exception("Unsupported Certificate source");
+        }
+
+        return certificates.Cast<X509Certificate>().ToArray();
+    }
+
+    private static X509Certificate2[] GetCertificatesFromString(string certificateContentsBase64, string keyPhrase)
+    {
+        var certificateBytes = Convert.FromBase64String(certificateContentsBase64);
+        return LoadCertificatesFromBytes(certificateBytes, keyPhrase);
+    }
+
+    // Have to skip these because there's no certificate to use for unit testing.
+    [ExcludeFromCodeCoverage]
+    private static X509Certificate2[] LoadCertificatesFromBytes(byte[] certificateBytes, string keyPhrase)
+    {
+        var collection = new X509Certificate2Collection();
+
+        if (!string.IsNullOrEmpty(keyPhrase))
+            collection.Import(certificateBytes, keyPhrase, X509KeyStorageFlags.PersistKeySet);
+        else
+            collection.Import(certificateBytes, null, X509KeyStorageFlags.PersistKeySet);
+
+        return collection.Cast<X509Certificate2>().OrderByDescending(c => c.HasPrivateKey).ToArray();
+
+    }
+
+    private static X509Certificate2[] GetCertificatesFromFile(string clientCertificateFilePath, string keyPhrase)
+    {
+        return LoadCertificatesFromBytes(File.ReadAllBytes(clientCertificateFilePath), keyPhrase);
+    }
+
+    // Have to skip these because there's no certificate to use for unit testing.
+    [ExcludeFromCodeCoverage]
+    private static X509Certificate2[] GetCertificatesFromStore(string thumbprint,
```

Comment:
You should be able to test this, since you can add and remove certificates from cert store: https://stackoverflow.com/a/567060 https://stackoverflow.com/a/7633249